server/src/main/java/io/druid/client/DirectDruidClient.java

```diff
@@ -168,16 +176,24 @@ public int getNumOpenConnections()
       final QueryMetrics<? super Query<T>> queryMetrics = toolChest.makeMetrics(query);
       queryMetrics.server(host);
 
+      long timeoutAt = ((Long) context.get(QUERY_FAIL_TIME)).longValue();
```

Comment:
It introduces NPE, if the query is made bypassing QueryResource, i. e. from inside QueryEngine of another type of query, via QuerySegmentWalker.